akka-actor/src/main/scala/akka/pattern/BackoffSupervisor.scala

```diff
@@ -404,11 +423,22 @@ private[akka] trait HandleBackoff { this: Actor ⇒
       context.parent ! msg
 
     case msg ⇒ child match {
-      case Some(c) ⇒ c.forward(msg)
-      case None ⇒ replyWhileStopped match {
-        case Some(r) ⇒ sender ! r
-        case None    ⇒ context.system.deadLetters.forward(msg)
-      }
+      case Some(c) ⇒
+        c.forward(msg)
+        if (!finalStopMessageReceived && finalStopMessage.isDefined) {
+          finalStopMessageReceived = finalStopMessage.get.apply(msg)
+        }
+      case None ⇒
+        replyWhileStopped match {
+          case None ⇒ context.system.deadLetters.forward(msg)
+          case _    ⇒ replyWhileStopped.foreach(a ⇒ sender() ! a)
+        }
+        if (finalStopMessage.isDefined) {
+          if (finalStopMessage.get.apply(msg)) {
```

Comment:
if the isDefined/get is because of avoiding allocations, I'd prefer pattern match